go/libraries/doltcore/diff/diffsplitter.go

```diff
@@ -40,128 +56,172 @@ type RowDiff struct {
+// |querySch| is the result schema from the dolt_dif(...) table function
+// it contains "from_..." and "to..." columns corresponding to the "from"
+// and "to" schemas used to generate the diff.
+// |targetSch| is the output schema used to print the diff and is computed
+// as the union schema of the "from" and "to" schemas.
+func NewDiffSplitter(querySch sql.Schema, targetSch sql.Schema) (*DiffSplitter, error) {
+	split, err := findDiffSchemaSplit(querySch)
+	if err != nil {
+		return nil, err
+	}
 
-		resultToTarget[i] = targetIdx
+	qtt, err := mapQuerySchemaToTargetSchema(querySch, targetSch)
+	if err != nil {
+		return nil, err
 	}
 
-	if fromLen == -1 {
-		fromLen = len(diffQuerySch) - 1
+	// not all "from..." columns will have a "to..." column and vice-versa
+	// unmapped columns in |qtq| are marked with -1
+	qtq, err := mapToAndFromColumns(querySch)
+	if err != nil {
+		return nil, err
 	}
 
 	return &DiffSplitter{
-		diffQuerySch:  diffQuerySch,
+		querySch:      querySch,
 		targetSch:     targetSch,
-		fromLen:       fromLen,
-		queryToTarget: resultToTarget,
-		fromTo:        fromTo,
-		toFrom:        toFrom,
+		queryToTarget: qtt,
+		queryToQuery:  qtq,
+		splitIdx:      split,
 	}, nil
 }
 
-func newRowDiff(size int) RowDiff {
-	return RowDiff{
-		ColDiffs: make([]ChangeType, size),
+func findDiffSchemaSplit(querySch sql.Schema) (int, error) {
+	split := -1
+	for i, col := range querySch {
+		switch {
+		case strings.HasPrefix(col.Name, fromPrefix):
+			if split >= 0 { // seen first "to_..." col
+				return 0, errors.New("interleaved 'from' and 'to' cols")
+			}
+		case strings.HasPrefix(col.Name, toPrefix):
+			if split < 0 { // |i| is first "to_..." col
+				split = i
+			}
+		case col.Name == "diff_type":
+			if split < 0 {
+				split = i
+			}
+		default:
+			return 0, errors.New("expected column prefix of 'to_' or 'from_' (" + col.Name + ")")
+		}
 	}
+	return split, nil
 }
 
-func (ds DiffSplitter) SplitDiffResultRow(row sql.Row) (RowDiff, RowDiff, error) {
-	// split rows in the result set into old, new
-	diffTypeColIdx := ds.diffQuerySch.IndexOfColName("diff_type")
-	if diffTypeColIdx < 0 {
-		return RowDiff{}, RowDiff{}, fmt.Errorf("expected a diff_type column")
+func mapQuerySchemaToTargetSchema(query, target sql.Schema) (mapping []int, err error) {
+	last := query[len(query)-1]
+	if last.Name != "diff_type" {
+		return nil, errors.New("expected last diff column to be 'diff_type'")
 	}
-
-	diffType := row[diffTypeColIdx]
-
-	oldRow, newRow := newRowDiff(len(ds.targetSch)), newRowDiff(len(ds.targetSch))
-
-	diffTypeStr := diffType.(string)
-	if diffTypeStr == "removed" || diffTypeStr == "modified" {
-		oldRow.Row = make(sql.Row, len(ds.targetSch))
-		if diffTypeStr == "modified" {
-			oldRow.RowDiff = ModifiedOld
-		} else {
-			oldRow.RowDiff = Removed
+	query = query[:len(query)-1]
+
+	mapping = make([]int, len(query))
+	for i, col := range query {
+		switch {
```

Comment:
Nit: this kind of conditional logic is better expressed as `if ... else if .. else`